### PR TITLE
[Relay][Module] Refactor the way we interface between different modules of Relay.

### PR DESCRIPTION
**Pull request by jroesch**
Introduces an idempotent import mechanism for Relay and restructures some of the code around handling Relay files. One of the challenges for introducing new dialects of Relay as I was doing in #3560 is the ability to define new external types, and operations over them. 

In order to track new types such as `Storage` we need the ability to share chunks of code in a idempotent way. The current module design is lacking in this regard and the import mechanism addresses this. 

An example of where this fails now is if we introduce new builtin types, where do we place the code to install them in a module? for example if we use the instantiate the prelude more than once we will receive an error right now. 

This is required for my memory analysis PR which is upcoming, and depends on @weberlo's parser PR.

@weberlo will follow up by moving the entire Prelude to the text format. 

This PR also cleans up InferType's handling of modules which is currently sub-optimal.

### CODE REVIEW

File: include/tvm/relay/module.h [hunk 1]
```diff
@@ -228,6 +245,12 @@ struct Module : public NodeRef {
   using ContainerType = ModuleNode;
 };
 
+/*! \brief Parse Relay source into a module.
+ * \param source A string of Relay source code.
+ * \param source_name The name of the source file.
+ * \return A Relay module.
+ */
+Module FromText(std::string source, const std::string& source_name);
```
Comment (MarisaKirisame) [inline include/tvm/relay/module.h]: const &

File: src/relay/ir/module.cc [hunk 2]
```diff
@@ -85,7 +89,9 @@ void ModuleNode::AddUnchecked(const GlobalVar& var,
 }
 
 GlobalTypeVar ModuleNode::GetGlobalTypeVar(const std::string& name) const {
+  CHECK(global_type_var_map_.defined());
   auto it = global_type_var_map_.find(name);
+  std::cout << "pass here";
```
Comment (MarisaKirisame) [inline src/relay/ir/module.cc]: cout

File: include/tvm/relay/expr.h [hunk 3]
```diff
@@ -575,6 +575,7 @@ std::string PrettyPrint(const NodeRef& node);
 std::string AsText(const NodeRef& node,
                    bool show_meta_data = true,
                    runtime::TypedPackedFunc<std::string(Expr)> annotate = nullptr);
+
```
Comment (MarisaKirisame) [inline include/tvm/relay/expr.h]: remove extra line

**Comment by zhiics** [comment]
Thanks everyone, this is now merged.

**Pull request merged**